### PR TITLE
Independent Bob: Retrieve modes, retaining CFrags, replete WorkOrders, and manual sanitization.

### LINKED ISSUE #1195: Provide path for Bob to create new WorkOrder for same Ursula for same Capsule
#874 empowered Bob to retrieve the same secret on multiple occasions by optionally caching `CFrags` (in fact, an entire activated `Capsule`).

### LINKED ISSUE #1197: If Bob is caching WorkOrders and CFrags, what to do with an unused WorkOrder?
Consider m,n==2,3.

Bob retrieves, making three WorkOrders.  The first two Ursulas he contacts are up and he is able to receive service and successfully activate the `Capsule` without needing to use the third WorkOrder.  What to do with it?

### LINKED ISSUE #1230: UrsulaConfiguration per se reads from raw.githubusercontent.com.
I only noticed this when I tried to run the test suite while offline.

```
ERROR
test setup failed
self = <urllib3.connection.VerifiedHTTPSConnection object at 0x7f3f69aedd68>

    def _new_conn(self):
        """ Establish a socket connection and set nodelay settings on it.
    ERROR
test setup failed
```
...
```
>           raise MaxRetryError(_pool, url, error or ResponseError(cause))
E           urllib3.exceptions.MaxRetryError: HTTPSConnectionPool(host='raw.githubusercontent.com', port=443): Max retries exceeded with url: /nucypher/ethereum-contract-registry/goerli/contract_registry.json (Caused by NewConnectionError('<urllib3.connection.VerifiedHTTPSConnection object at 0x7f3f69aedd68>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution'))

../../../../.virtualenvs/nucypher-EzeC4Bqx/lib/python3.7/site-packages/urllib3/util/retry.py:399: MaxRetryError

During handling of the above exception, another exception occurred:

    @pytest.fixture(scope="module")
    def ursula_federated_test_config():
        ursula_config = UrsulaConfiguration(dev_mode=True,
                                            rest_port=MOCK_URSULA_STARTING_PORT,
                                            start_learning_now=False,
                                            abort_on_learning_error=True,
                                            federated_only=True,
                                            network_middleware=MockRestMiddleware(),
                                            save_metadata=False,
>                                           reload_metadata=False)

../fixtures.py:132: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../../nucypher/config/characters.py:69: in __init__
    super().__init__(dev_mode=dev_mode, *args, **kwargs)
../../nucypher/config/node.py:154: in __init__
    self.initialize(password=DEVELOPMENT_CONFIGURATION)
../../nucypher/config/node.py:394: in initialize
    self.registry_filepath = EthereumContractRegistry.download_latest_publication()
../../nucypher/blockchain/eth/registry.py:94: in download_latest_publication
    response = requests.get(github_endpoint)
../../../../.virtualenvs/nucypher-EzeC4Bqx/lib/python3.7/site-packages/requests/api.py:75: in get
    return request('get', url, params=params, **kwargs)
../../../../.virtualenvs/nucypher-EzeC4Bqx/lib/python3.7/site-packages/requests/api.py:60: in request
    return session.request(method=method, url=url, **kwargs)
../../../../.virtualenvs/nucypher-EzeC4Bqx/lib/python3.7/site-packages/requests/sessions.py:533: in request
    resp = self.send(prep, **send_kwargs)
../../../../.virtualenvs/nucypher-EzeC4Bqx/lib/python3.7/site-packages/requests/sessions.py:646: in send
    r = adapter.send(request, **kwargs)
```
...
```
>           raise ConnectionError(e, request=request)
E           requests.exceptions.ConnectionError: HTTPSConnectionPool(host='raw.githubusercontent.com', port=443): Max retries exceeded with url: /nucypher/ethereum-contract-registry/goerli/contract_registry.json (Caused by NewConnectionError('<urllib3.connection.VerifiedHTTPSConnection object at 0x7f3f69aedd68>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution'))

../../../../.virtualenvs/nucypher-EzeC4Bqx/lib/python3.7/site-packages/requests/adapters.py:516: ConnectionError
```

### LINKED ISSUE #1346: Mocking LocalContractRegistry.read causes strange blockchain-related failures in subsequent tests.
See, for example:

https://app.circleci.com/jobs/github/nucypher/nucypher/47088/tests

### LINKED ISSUE #1359: Work Order can have more than one Capsule.  Deal with it.  (Part I: What About Bob?)
Until now, we have always assumed that Bob's adventure has him slinging a single `Capsule` under his arm.  In a few places in the codebase, we have noted that this is not sensible and vowed (well, `TODO`'d) to do something about it.  Some examples include `retrieve()`, `generate_work_orders()`, and `WorkOrder.construct_by_bob()`.

These areas need to account for Bob having more than one (perhaps a huge sack full) of Capsules.

### LINKED ISSUE #1362: CryptoKits need to use BytestringKwargifier unless there's an articulated reason (in the docstring) not to.

### PR DESCRIPTION
**Pull request by jMyles**
This PR is designed to bring Bob up to a minimum-viable level functionality in line with the other Characters.

In short: it makes most ordinary Bob actions just work, whereas before they needed workarounds.

## Retained CFrags, replete WorkOrders
Bob can now optionally retain attached CFrags on a Capsule.  Bob can also retain "replete" WorkOrders, which have already been completed by Ursula for all Capsules in the WorkOrder.
Fixes #1197: Unused WorkOrders are no longer retained.

## Multiple Capsules
Fixes #1359: All retrieval logic now accounts for scenarios in which more than one Capsule is being activated in a single adventure.
Fixes #1195: Bob can re-issue a WorkOrder by simply removing the old one and retrieving again.

## Bugfixes
Fixes #1362: Kits are more now more cleanly de/serialized and use `BytestringKwargifier`.  Also, splitters for kit constituents are reused in kit splitters.
Fixes #1346: Blockchain mocks now occur in a try-finally block.
Fixes #1230.

Left undone:
#1248 and #1249: Higher-level tuning knobs such as CLI args and config items need to be put in place.
Docs: The docs are just stubs.  Much more to be done.

### CODE REVIEW

**Comment by KPrasch** [comment]
Do you want review on this PR? I noticed it is not marked *WIP*, and also not going into an EPIC.

**Comment by michwill** [comment]
I think @jMyles gave a lot of effort to this PR, so it's probably not good to leave it waiting in the dust

**Comment by jMyles** [comment]
@KPrasch: I'm rebasing it now to go into `birdcage`.  Just dealing with a few nagging tests at this point.  Hopefully tonight!

File: docs/source/guides/retrieving_secrets_with_bob.rst [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+Basic retrieval
+---------------
+
+Alice has created a Policy and Granted Bob access.
+Enrico has given Bob a Capsule.
+Now Bob wants the message.
+
+```
+retrieve()
```
Comment (michwill) [inline docs/source/guides/retrieving_secrets_with_bob.rst]: That's a very fun guide. Am i right that it's not intended to tell how would *exactly* Bob retrieve?

Comment (michwill) [reply docs/source/guides/retrieving_secrets_with_bob.rst]: (e.g. what arguments to use, a self-sufficient example, ...)

File: nucypher/policy/collections.py [hunk 2]
```diff
@@ -290,15 +291,15 @@ def construct_by_bob(cls, arrangement_id, capsules, ursula, bob):
         if ursula._stamp_has_valid_signature_by_worker():
             ursula_identity_evidence = ursula.decentralized_identity_evidence
 
-        tasks, tasks_bytes = [], []
+        tasks, tasks_bytes = {}, []
```
Comment (michwill) [inline nucypher/policy/collections.py]: Is it important for `tasks_bytes` to have the same order as tasks? (I understand that in Python 3.7+ it'll be the case anyway)

**Comment by jMyles** [comment]
@fjarri: No specific reason; I just re-ran lock and the tests passed with these (slightly updated versions).  So, it was prompted by BSS, but I figured a general update was in order as long as things pass.

@mswilkison: Oh agreed of course; the docs are far from finished.  But I'm surprised to hear that you want these removed and added instead to a follow-up PR.  We have often had epics with docs-in-progress.  Are you wanting to change that practice?  As for "replete" - yeah it's a bit of an esoteric term, but the thesaurus isn't helping.  I'm having some brainblock on how to improve it.

File: docs/source/guides/retrieving_secrets_with_bob.rst (hunk 1 above)
Comment (jMyles) [reply docs/source/guides/retrieving_secrets_with_bob.rst]: Yeah, as we talked about in Osaka, a subsequent PR in this same epic will advance the docs.

File: nucypher/policy/collections.py (hunk 2 above)
Comment (jMyles) [reply nucypher/policy/collections.py]: :heavy_check_mark:

**Comment by mswilkison** [comment]
> @mswilkison: Oh agreed of course; the docs are far from finished. But I'm surprised to hear that you want these removed and added instead to a follow-up PR. We have often had epics with docs-in-progress. Are you wanting to change that practice? As for "replete" - yeah it's a bit of an esoteric term, but the thesaurus isn't helping. I'm having some brainblock on how to improve it.

If we have a follow-up PR into the epic before master it's fine!

File: tests/cli/test_bob.py [hunk 3]
```diff
@@ -8,3 +23,93 @@ def test_bob_cannot_init_with_dev_flag(click_runner):
...
+
+
+    _old_make_character_function = actions.make_cli_character
+    try:
+
+        log.info("Patching make_cli_character with substitute_bob")
+        actions.make_cli_character = substitute_bob
+
+        # Once...
+        with GlobalLoggerSettings.pause_console_logging_while():
```
Comment (michwill) [inline tests/cli/test_bob.py]: Oh wow, didn't know you can do it that way!

File: nucypher/characters/lawful.py [hunk 4]
```diff
@@ -1299,6 +1399,12 @@ def from_alice(cls, alice: Alice, label: bytes):
         return cls(crypto_power_ups={SigningPower: alice.stamp.as_umbral_pubkey()},
                    policy_encrypting_key=policy_pubkey_enc)
 
+    @property
+    def policy_pubkey(self):
+        if not self._policy_pubkey:
+            raise TypeError("This Enrico doesn't know which policy encrypting key he used.  Oh well.")
```
Comment (derekpierre) [inline nucypher/characters/lawful.py]: Haha is it "Oh well" if we are raising an exception about it...?

**Pull request merged**